### DISCUSSION TITLE
psycopg3 equivalent of IN with tuple of tuples

### DISCUSSION THREAD

**Discussion by henribru** [psycopg/psycopg#486] [Q&A]
I have some psycopg2 code which does essentially `conn.execute("SELECT * FROM mytable WHERE (a, b) IN %s", ((1, 2), (3, 4)))`. As mentioned in https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#you-cannot-use-in-s-with-a-tuple you can't use `IN` with a tuple in psycopg3 but should use `ANY` with an array instead. My attempt at this was `conn.execute("SELECT * FROM mytable WHERE (a, b) = ANY(%s)", [[1, 2], [3, 4]])`, but this leads to Postgresql giving the error "input of anonymous composite types is not implemented". Does this mean the proper way to translate this code to psycopg2 is to register a new composite type or is there a simpler workaround?

**Comment by dvarrazzo** [comment]
Your example might have a problem, because you are passing two arguments to the query, and I get the error "the query has 1 placeholders but 2 parameters were passed".

Fixed that obvious one, I get a different error:

```
conn.execute("SELECT * FROM mytable WHERE (a, b) = ANY(%s)", [[[1, 2], [3, 4]]])

UndefinedFunction: operator does not exist: record = smallint
LINE 1: SELECT * FROM mytable WHERE (a, b) = ANY($1)
                                           ^
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```

It seems that, when postgres receives a 2d array in ANY, it unrolls it completely. Same thing that I can see happening in:

```
piro=# select unnest('{{1,2},{3,4}}'::smallint[]);
┌────────┐
│ unnest │
├────────┤
│      1 │
│      2 │
│      3 │
│      4 │
└────────┘
(4 rows)
```

And, trying to be as explicit as possible, leads to the same result:

```
=# SELECT * FROM mytable WHERE (a,b) = ANY(array[array[1,2],array[3,4]]);
ERROR:  operator does not exist: record = integer

=# SELECT * FROM mytable WHERE array[a,b] = ANY(array[array[1,2],array[3,4]]);
ERROR:  operator does not exist: integer[] = integer
```

this works instead:

```
SELECT * FROM mytable WHERE (a,b) = ANY(array[row(1,2),row(3,4)]);
```

so we need a way to pass the oid of the array of the type record to postgres. Psycopg doesn't have a dumper for tuples so we can use that. I tried this (just because the Python repr of a tuple of integers is valid as record format, but this is not safe or generic):

```python
from typing import Tuple, Any

import psycopg
from psycopg.adapt import Dumper

conn = psycopg.connect()


class TupleDumper(Dumper):
    oid = psycopg.postgres.types["record"].oid

    def dump(self, obj: Tuple[Any, ...]) -> bytes:
        return str(obj).encode()


conn.adapters.register_dumper(tuple, TupleDumper)

cur = conn.execute("SELECT * FROM mytable WHERE (a,b) = ANY(%s)", [[(1, 2), (3, 4)]])
print(cur.fetchall())
```

and unfortunately this has the same problem :\

```
(.venv) piro@sherekhan:~/dev/psycopg3$ python test_486.py 
Traceback (most recent call last):
  File "/home/piro/dev/psycopg3/test_486.py", line 18, in <module>
    cur = conn.execute("SELECT * FROM mytable WHERE (a,b) = ANY(%s)", [[(1, 2), (3, 4)]])
  File "/home/piro/dev/psycopg3/psycopg/build/__editable__.psycopg-3.2.0.dev1-py3-none-any/psycopg/connection.py", line 879, in execute
    raise ex.with_traceback(None)
psycopg.errors.FeatureNotSupported: input of anonymous composite types is not implemented
CONTEXT:  unnamed portal parameter $1 = '...'
```

This works instead, but yes, creating a composite type:

```python
from typing import Tuple, Any

import psycopg
from psycopg.adapt import Dumper

conn = psycopg.connect()

try:
    with conn.transaction():
        conn.execute("create type intint as (a integer, b integer)")
except psycopg.errors.DuplicateObject:
    pass

t: psycopg.types.TypeInfo
t = psycopg.types.TypeInfo.fetch(conn, "intint")  # type: ignore
assert t
conn.adapters.types.add(t)


class TupleDumper(Dumper):
    oid = t.oid

    def dump(self, obj: Tuple[Any, ...]) -> bytes:
        return str(obj).encode()


conn.adapters.register_dumper(tuple, TupleDumper)

cur = conn.execute("SELECT * FROM mytable WHERE (a,b) = ANY(%s)", [[(1, 2), (3, 4)]])
print(cur.fetchall())
```

We can add support in psycopg to make this easier, but it would still require to create an ad-hoc composite, and because its oid would be unknown we need to fetch it.

Let me ask a postgres mailing list if there is a better workaround.

**Reply by henribru** [reply]
Apologies for the poor state of my reproduction, was juggling too many attempts and copy-pasted the wrong one. It was actually `execute("SELECT * FROM mytable WHERE (a,b) = ANY(%s)", [[(1, 2), (3, 4)]])` that gave the error, like in your reproduction. I didn't have to register a `TupleDumper` first, but I was running this indirectly through Django and perhaps Django registers something equivalent. I should have mentioned that part (or just tried it directly with psycopg), apologies again, I mistakenly assumed it would be passed through to psycopg without much modification.

**Reply by dvarrazzo** [reply]
No worries. psycopg has a problem with this pattern indeed. It would be nice to provide a better solution than the string composition mentioned in the other answer.

**Comment by dvarrazzo** [comment]
It is possible to use the syntax we found working using client-side composition of the argument. This is type-safe instead:

```python
from typing import Any, Sequence, Tuple

import psycopg
from psycopg import sql

conn = psycopg.connect()


def array_of_rows(items: Sequence[Tuple[Any, ...]]) -> sql.Composable:
    rows = [
        sql.SQL("row(")
        + sql.SQL(",").join([sql.Literal(v) for v in item])
        + sql.SQL(")")
        for item in items
    ]

    return sql.SQL("array[") + sql.SQL(",").join(rows) + sql.SQL("]")


cur = conn.execute(
    sql.SQL("SELECT * FROM mytable WHERE (a,b) = ANY({})").format(
        array_of_rows([(1, 2), (3, 4)])
    )
)

print(cur.fetchall())
```
